nri-kafka/src/Kafka/Worker.hs

```diff
@@ -0,0 +1,505 @@
+{-# LANGUAGE GADTs #-}
+
+-- The Kafka worker has the following concurrent workflows:
+-- 1. The main thread which handles
+--   - gracefully quitting
+--   - a single-thread `pollingLoop` that reads new messages from kafka
+-- 2. The Kafka `rebalanceCallback` which handles rebalancing partitions, in turn
+--    turning on and off worker threads.
+-- 3. A single-thread `pauseAndAnalyticsLoop` that tells the kafka library to pause & resume
+--    sending us messages from specific partitions (based on a number of factors)
+-- 4. Multiple worker-threads (`Partition.processMsgLoop`), one per
+--    (topic,partition)
+
+-- | Kafka.Worker is a module for processing a Kafka log.
+-- It can be used to build a CLI that will consume and process a user-defined message type
+module Kafka.Worker
+  ( process,
+    Description (..),
+    Partition.MessageCallback (Partition.MessageCallback),
+    OffsetSource (..),
+    Consumer.ConsumerGroupId (..),
+    PartitionKey,
+    Consumer.TopicName (..),
+    Consumer.ConsumerRecord (..),
+    Partition.ProcessResult (..),
+    Consumer.PartitionId (..),
```

Comment:
:art: Lots of wrapper types here. I think it makes sense to have these inside the library, but wonder if we can tweak our API's so library users don't need to deal with them as much. Going to check.